### PR TITLE
add FallbackMailer

### PR DESCRIPTION
**Pull request by JanTvrdik**
Simple mailer which can retry sending mail multiple times with either one ore multiple mailers.

### CODE REVIEW

**Comment by JanTvrdik** [comment]
Some open questions:

* Should the total number of retries by `$retryCount` or `$retryCount * count($this->mailers)`?
* Should we log the exceptions? I don't know how to do it without depending on Tracy\ILogger. Or should we just emit warnings?

**Comment by dg** [comment]
To throw, to log is task for a diffent layer.

**Comment by JanTvrdik** [comment]
Poetic but I don't understand what you mean. Does that mean that I should move the FallbackMailer because it cannot be implemented in Nette?

**Comment by dg** [comment]
I tried to say that FallbackMailer should not log exceptions. It is not its responsibility. It can provide some callback, but not log them.

File: src/Mail/FallbackMailer.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+
+	/**
+	 * Sends email.
+	 * @return void
+	 * @throws SendException
+	 */
+	public function send(Message $mail)
+	{
+		for ($i = 0; $i < $this->retryCount; $i++) {
```
Comment (milo) [inline src/Mail/FallbackMailer.php]: What about to check, that $mailers is not empty?

Comment (JanTvrdik) [inline src/Mail/FallbackMailer.php]: Good point, I'll add the check in constructor.

File: src/Mail/FallbackMailer.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+					$mailer->send($mail);
+					return;
+
+				} catch (SendException $e) {
+					$this->onFailure($this, $e, $mailer, $mail);
+				}
+			}
+		}
+
+		throw $e;
```
Comment (JanTvrdik) [inline src/Mail/FallbackMailer.php]: self note: wrap this with another SendException instance

**Comment by JanTvrdik** [comment]
This is ready to merge from my point of view.

File: src/Mail/FallbackMailer.php [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	public function getRetryCount()
+	{
+		return $this->retryCount;
+	}
+
+
+	/**
+	 * @param int
+	 */
+	public function setRetryCount($retryCount)
```
Comment (dg) [inline src/Mail/FallbackMailer.php]: All setters in nette should return `$this`. But even better is to omit anything that is not really needed. I think that except `addMailer()` is not need nothing. And with `addMailer()` should be `if (!$mailers) { throw ...` tested later.

**Comment by f3l1x** [comment]
Wow. Good job. :+1: 

-----

I was not sure it could be part of nette/mail package. I use for every project on my local devstack something that I called FileMailer. Do you think it could be also part of package?

Implementation is really small and cool.

```php
<?php
use Nette\Mail\IMailer;
use Nette\Mail\Message;

final class FileMailer implements IMailer
{

    /** @var string */
    private $path;

    /**
     * @param string $path
     */
    public function __construct($path)
    {
        $this->path = $path;
    }

    /**
     * @param Message $mail
     */
    public function send(Message $mail)
    {
        @mkdir($path, 0777, TRUE);
        file_put_contents($this->path . date('Y-m-d H-i-s') . microtime() . '.eml', $mail->generateMessage());
    }

}
```

Should I prepare PR?

**Comment by JanTvrdik** [comment]
@f3l1x What do you use it for?

**Comment by f3l1x** [comment]
Dump message to folder, open in editor, check design especially for newsletter, verify all EML options/requirements (html vs plain) etc.

**Pull request closed without merging**